meegkit/trca.py

```diff
@@ -210,9 +179,154 @@ def predict(self, X):
+    S = np.zeros((n_chans, n_chans))
+    for trial_i in range(n_trials - 1):
+        x1 = np.squeeze(X[..., trial_i])
+
+        # Mean centering for the selected trial
+        x1 -= np.mean(x1, 0)
+
+        # Select a second trial that is different
+        for trial_j in range(trial_i + 1, n_trials):
+            x2 = np.squeeze(X[..., trial_j])
+
+            # Mean centering for the selected trial
+            x2 -= np.mean(x2, 0)
+
+            # Compute empirical covariance between the two selected trials and
+            # sum it
+            S = S + x1.T @ x2 + x2.T @ x1
+
+    # 3. Compute eigenvalues and vectors
+    # -------------------------------------------------------------------------
+    lambdas, W = linalg.eig(S, Q, left=True, right=False)
+
+    # Select the eigenvector corresponding to the biggest eigenvalue
+    W_best = W[:, np.argmax(lambdas)]
+
+    return W_best
+
+
+def trca_regul(X, method):
+    """Task-related component analysis.
+
+    This function implements a variation of the method described in [1]_. It is
+    inspired by a riemannian geometry approach to CSP [2]_. It adds
+    regularization to the covariance matrices and uses the riemannian mean for
+    the inter-trial covariance matrix `S`.
+
+    Parameters
+    ----------
+    X : array, shape=(n_samples, n_chans[, n_trials])
+        Training data.
+
+    Returns
+    -------
+    W : array, shape=(n_chans,)
+        Weight coefficients for electrodes which can be used as a spatial
+        filter.
+
+    References
+    ----------
+    .. [1] M. Nakanishi, Y. Wang, X. Chen, Y. -T. Wang, X. Gao, and T.-P. Jung,
+       "Enhancing detection of SSVEPs for a high-speed brain speller using
+       task-related component analysis", IEEE Trans. Biomed. Eng,
+       65(1):104-112, 2018.
+    .. [2] Barachant, A., Bonnet, S., Congedo, M., & Jutten, C. (2010,
+       October). Common spatial pattern revisited by Riemannian geometry. In
+       2010 IEEE International Workshop on Multimedia Signal Processing (pp.
+       472-476). IEEE.
+
+    """
+    n_samples, n_chans, n_trials = theshapeof(X)
+
+    # 1. Compute empirical covariance of all data (to be bounded)
+    # -------------------------------------------------------------------------
+    # Concatenate all the trials to have all the data as a sequence
+    UX = np.zeros((n_chans, n_samples * n_trials))
+    for trial in range(n_trials):
+        UX[:, trial * n_samples:(trial + 1) * n_samples] = X[..., trial].T
+
+    # Mean centering
+    UX -= np.mean(UX, 1)[:, None]
+
+    # Compute empirical variance of all data (to be bounded)
+    cov = Covariances(estimator=method).fit_transform(UX[np.newaxis, ...])
+    Q = np.squeeze(cov)
+
+    # 2. Compute average empirical covariance between all pairs of trials
+    # -------------------------------------------------------------------------
+    # Intertrial correlation computation
+    data = np.concatenate((X, X), axis=1)
```

Comment:
should the data not me zero-centered here before computing the covariance?

Reply:
Actually, `pyriemann.estimation.Covariances` [rely](https://github.com/alexandrebarachant/pyRiemann/blob/master/pyriemann/utils/covariance.py) on `sklearn.covariance` and by default `assume_centered` is set to `False` (in `schaefer_strimmer_cov` function the data are zero-centered as well).